context/app/utils.py

```diff
@@ -62,3 +62,49 @@ def get_organs():
     dir_path = Path(dirname(__file__) + '/organ')
     organs = {p.stem: safe_load(p.read_text()) for p in dir_path.glob('*.yaml')}
     return organs
+
+
+# Redirect to primary dataset if this entity is
+# - non-existent
+# - a support entity (e.g. an image pyramid)
+# - a processed or component dataset
+def should_redirect_entity(entity):
+    if not entity:
+        return True
+
+    actual_type = entity.get('entity_type').lower()
+    is_support_type = actual_type == 'support'
+    is_component = entity.get('is_component', False) is True
+    is_not_raw_dataset = entity.get('processing') != 'raw' and actual_type == 'dataset'
+
+    if is_support_type or is_component or is_not_raw_dataset:
+        return True
+
+    return False
+
+
+def find_earliest_dataset_ancestor(client, uuid):
+    dataset = client.get_entities(
+        'datasets',
+        query_override={
+            "bool": {
+                "must": {
+                    "term": {
+                        "uuid": uuid
+                    }
+                }
+            }
+        },
+        non_metadata_fields=['hubmap_id', 'uuid', 'immediate_ancestors', 'entity_type']
+    )
+
+    # If no dataset is found or it has no ancestors, return the current dataset UUID
+    if not dataset or not dataset[0].get('immediate_ancestors'):
+        return uuid
+
+    # Traverse through immediate ancestors to find the earliest dataset ancestor
+    for ancestor in dataset[0]['immediate_ancestors']:
+        if ancestor.get('entity_type') == 'Dataset':
+            uuid = find_earliest_dataset_ancestor(client, ancestor.get('uuid'))
```

Comment:
We shouldn't depend on `immediate_ancestors` since it's somewhat deprecated and we should also avoid recursion and multiple requests if possible. 

Can we use a query that returns the dataset which has `processing === 'raw'`, does not have the field `ancestor_counts.entity_type.Dataset`, and its `uuid` is included in the given dataset's `ancestors_ids`? Let me know if you need help drafting the query.